src/rt/tracegc.d

```diff
@@ -88,243 +84,156 @@ extern (C) void[] _d_newarrayT(const TypeInfo ti, size_t length);
+            name.length, name.ptr
+        );
+    }
 
+    auto stats1 = GC.stats();
+    scope(exit)
+    {
+        auto stats2 = GC.stats();
+        if (stats2.totalCollected < stats1.totalCollected)
+        {
+            // need to account for unsigned overflow possibility if app is being
+            // run very long
+            stats2.totalCollected += typeof(stats1.totalCollected).max
+                - stats1.totalCollected;
+            stats1.totalCollected = 0;
+        }
+        ulong size = (stats2.usedSize + stats2.totalCollected) -
+            (stats1.usedSize + stats1.totalCollected);
+        if (size > 0)
+            accumulate(file, line, funcname, name, size);
+    }
 
+    return original_func();
+};
 
-extern (C) byte[] _d_arraycatT(const TypeInfo ti, byte[] x, byte[] y);
-extern (C) void[] _d_arraycatnTX(const TypeInfo ti, byte[][] arrs);
+mixin(generateTraceWrappers());
+//pragma(msg, generateTraceWrappers());
 
-extern (C) byte[] _d_arraycatTTrace(string file, int line, string funcname, const TypeInfo ti, byte[] x, byte[] y)
-{
-    mixin(generatePrintf());
-    accumulate(file, line, funcname, ti.toString(), (x.length + y.length) * ti.next.tsize);
-    return _d_arraycatT(ti, x, y);
-}
+////////////////////////////////////////////////////////////////////////////////
+// code gen implementation
 
-extern (C) void[] _d_arraycatnTXTrace(string file, int line, string funcname, const TypeInfo ti, byte[][] arrs)
+private string generateTraceWrappers()
 {
-    mixin(generatePrintf());
-    size_t length;
-    foreach (b; arrs)
-        length += b.length;
-    accumulate(file, line, funcname, ti.toString(), length * ti.next.tsize);
-    return _d_arraycatnTX(ti, arrs);
-}
+    string code;
 
-extern (C) void[] _d_arrayappendT(const TypeInfo ti, ref byte[] x, byte[] y);
-extern (C) byte[] _d_arrayappendcTX(const TypeInfo ti, ref byte[] px, size_t n);
-extern (C) void[] _d_arrayappendcd(ref byte[] x, dchar c);
-extern (C) void[] _d_arrayappendwd(ref byte[] x, dchar c);
+    foreach (name; __traits(allMembers, mixin(__MODULE__)))
+    {
+        static if (name.length > 3 && name[0..3] == "_d_")
+        {
+            mixin("alias Declaration = " ~ name ~ ";");
+            code ~= generateWrapper!Declaration();
+        }
+    }
 
-extern (C) void[] _d_arrayappendTTrace(string file, int line, string funcname, const TypeInfo ti, ref byte[] x, byte[] y)
-{
-    mixin(generatePrintf());
-    accumulate(file, line, funcname, ti.toString(), ti.next.tsize * y.length);
-    return _d_arrayappendT(ti, x, y);
+    return code;
 }
 
-extern (C) byte[] _d_arrayappendcTXTrace(string file, int line, string funcname, const TypeInfo ti, ref byte[] px, size_t n)
+private string generateWrapper(alias Declaration)()
 {
-    mixin(generatePrintf());
-    accumulate(file, line, funcname, ti.toString(), ti.next.tsize * n);
-    return _d_arrayappendcTX(ti, px, n);
-}
+    static size_t findParamIndex(string s)
```

Comment:
Would be very nice if these helper methods would have ddoc, or at least `generateWrapper`.